### PR TITLE
Add API for disallowing writes to a directory

### LINKED ISSUE #36: Add ability to reset permissions on a folder to read-only
Once accessioning has started on files in a directory, we need to set permissions on that directory to read-only so that the user can no longer add or change files.

### PR DESCRIPTION
**Pull request by mjgiarlo**
Fixes #36

# Why was this change made? 🤔

This commit adds the ability to disallow writes to a directory (i.e., make a directory read-only) to the public API. As part of this work, add a minor refactoring of the access-related components within the `Endpoint` class for greater code re-use. Also includes:
* Establish a new pattern for streamlined testing of `Client` public API
* Rename `Endpoint#set_permissions` to `Endpoint#allow_writes`
* Allow access changes to handle rule creation and rule updates
* Add and document an integration test for devs.

# How was this change tested? 🤨

CI and ran the integration test

### CODE REVIEW

File: lib/globus/client/endpoint.rb [hunk 1]
```diff
@@ -84,8 +69,8 @@ def connection
         )
       end
 
-      def identity
-        Globus::Client::Identity.new(config)
+      def user
+        Identity.new(config).get_identity_id(user_id)
       end
```
Comment (mjgiarlo) [inline lib/globus/client/endpoint.rb:74]: This class doesn't really care about identities; it just needs a user to fill out the `principal` field within an access request.

File: lib/globus/client/endpoint.rb [hunk 2]
```diff
@@ -98,29 +83,58 @@ def paths
         end
       end
 
+      # @see #paths
+      def full_path
+        paths.last
+      end
```
Comment (mjgiarlo) [inline lib/globus/client/endpoint.rb:89]: `paths.last` was showing up in three places, so I figured it was time to give it a name.

File: lib/globus/client/endpoint.rb [hunk 3]
```diff
@@ -98,29 +83,58 @@ def paths
...
-        response = connection.get("#{transfer_path}/ls?path=#{paths.last}")
+        response = connection.get("#{transfer_path}/ls?path=#{full_path}")
         return JSON.parse(response.body) if response.success?
 
         UnexpectedResponse.call(response)
       end
 
+      def files
+        objects["DATA"].select { |object| object["DATA_TYPE"] == "file" }
+      end
```
Comment (mjgiarlo) [inline lib/globus/client/endpoint.rb:129]: Moved this up so `files` and `objects` could be buddies, and the `_path` methods below could be buddies.

File: lib/globus/client/endpoint.rb [hunk 4]
```diff
@@ -98,18 +83,69 @@ def paths
...
+              notify_email: "#{user_id}@stanford.edu"
+            }.to_json
+            req.headers["Content-Type"] = "application/json"
+          end
+        end
+
+        return response if response.success?
+
+        UnexpectedResponse.call(response)
+      end
```
Comment (mjgiarlo) [inline lib/globus/client/endpoint.rb:133]: Now all access/permissions changes can handle updates as well as creates; we no longer assume that first we set to read-write and then to read-only.

File: lib/globus/client/endpoint.rb (hunk 2 above)
Comment (lwrubel) [reply lib/globus/client/endpoint.rb:89]: New name is clearer, too.

**Review by lwrubel — APPROVED** [review]

**Pull request merged**